File: 51-n-queens/51-n-queens.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> result;
        string row;
        for(int i = 0; i < n; i++)
            row.push_back('.');
        vector<string> board;
        for(int i = 0; i < n; i++)
            board.push_back(row);
        solve(0, n, board, result);
        return result;
    }
    
    void solve(int col, int n, vector<string> &board, vector<vector<string>> &ans) {
        if(col == n) {
            ans.push_back(board);
            return;
        }
        
        for(int row = 0; row < n; row++) {
            if(isSafe(row, col, n, board)) {
                board[row][col] = 'Q';
                solve(col + 1, n, board, ans);
                board[row][col] = '.';
            }
        }
    }
    
    bool isSafe(int row, int col, int n, vector<string> &board) {
        for(int j = 0; j < col; j++){
            if(board[row][j] == 'Q') {
                return false;
            }
        }
        
        for(int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if(board[i][j] == 'Q') {
                return false;
            }
        }
        
        for(int i = row, j = col; i < n && j >= 0; i++, j--) {
            if(board[i][j] == 'Q') {
                return false;
            }
        }
        
        return true;
    }
};

File: 51-n-queens/51-n-queens.py
class Solution:    
    
    def solveNQueens(self, n: int) -> List[List[str]]:
        board = []
        for i in range(n):
            row = []
            for j in range(n):
                row.append(".")
            board.append(row)
        
        result = []
    
        def is_safe(row, col, n, board):
            # print(row, col)
            for j in range(col):
                if board[row][j] == "Q":
                    return False

            i, j = row, col
            while i >= 0 and j >= 0:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j -= 1

            i, j = row, col
            while i < n and j >= 0:
                if board[i][j] == "Q":
                    return False
                i += 1
                j -= 1

            return True

        def solve(col, n, board, result):
            def display_board(board):
                for row in board:
                    print(row)
            
            if col == n:
                # display_board(board)
                # print()
                curr = [''.join(r) for r in board]
                result.append(curr)
                return

            # print(col)
            for row in range(n):
                if is_safe(row, col, n, board):
                    board[row][col] = "Q"
                    # display_board(board)
                    # print()
                    solve(col+1, n, board, result)
                    board[row][col] = "."
        
        # print("Initial Board")
        
        
        # display_board(board)
        
        solve(0, n, board, result)
        return result